topdown/http_test.go

```diff
@@ -1464,6 +1464,141 @@ func TestHTTPSendInterQueryForceCaching(t *testing.T) {
+		response         string
+		expectedReqCount int
+	}{
+		{
+			note:             "http.send GET cache expired, reloads normally",
+			request:          `{"method": "get", "url": "%URL%", "force_json_decode": true, "force_cache": true, "force_cache_duration_seconds": %CACHE%}`,
+			headers:          map[string][]string{},
+			expectedReqCount: 2,
+			response:         `{"x": 1}`,
+		},
+		{
+			note:             "http.send GET cache expired, no date, reloads normally",
+			request:          `{"method": "get", "url": "%URL%", "force_json_decode": true, "force_cache": true, "force_cache_duration_seconds": %CACHE%}`,
+			headers:          map[string][]string{},
+			expectedReqCount: 2,
+			skipDate:         true,
+			response:         `{"x": 1}`,
+		},
+		{
+			note:             "http.send GET cache expired, returns not modified",
+			request:          `{"method": "get", "url": "%URL%", "force_json_decode": true, "force_cache": true, "force_cache_duration_seconds": %CACHE%}`,
+			headers:          map[string][]string{"Etag": {"1234"}},
+			expectedReqCount: 2,
+			response:         `{"x": 1}`,
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.note, func(t *testing.T) {
+			t0 := time.Now().UTC()
+
+			var requests []*http.Request
+			ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				requests = append(requests, r)
+				headers := w.Header()
+
+				for k, v := range tc.headers {
+					headers[k] = v
+				}
+
+				if tc.skipDate {
+					headers["Date"] = nil
+				} else {
+					headers.Set("Date", t0.Format(http.TimeFormat))
+				}
+
+				etag := w.Header().Get("etag")
+
+				if r.Header.Get("if-none-match") != "" {
+					if r.Header.Get("if-none-match") == etag {
+						// add new headers and update existing header value
+						headers["Cache-Control"] = []string{"max-age=200, public"}
+						w.WriteHeader(http.StatusNotModified)
+					}
+				} else {
+					w.WriteHeader(http.StatusOK)
+					_, err := w.Write([]byte(tc.response))
+					if err != nil {
+						t.Fatal(err)
+					}
+				}
+			}))
+			defer ts.Close()
+
+			request := strings.ReplaceAll(tc.request, "%URL%", ts.URL)
+			request = strings.ReplaceAll(request, "%CACHE%", strconv.Itoa(cacheTime))
+			full := fmt.Sprintf("http.send(%s, x)", request)
+			config, _ := iCache.ParseCachingConfig(nil)
+			interQueryCache := iCache.NewInterQueryCache(config)
+			q := NewQuery(ast.MustParseBody(full)).
+				WithInterQueryBuiltinCache(interQueryCache).
+				WithTime(t0)
+
+			for i := 0; i < 2; i++ {
+				resp, err := q.Run(context.Background())
+				if err != nil {
+					t.Fatal(err)
+				}
+
+				if len(resp) < 1 {
+					t.Fatalf("missing response on query %d: %v", i, resp)
+				}
+
+				resResponse := resp[0]["x"].Value.(ast.Object).Get(ast.StringTerm("raw_body"))
+				if ast.String(tc.response).Compare(resResponse.Value) != 0 {
+					t.Fatalf("Expected response on query %d to be %v, got %v", i, tc.response, resResponse.String())
+				}
+
```

Comment:
The idea behind the logic below seems to simulate an expired cache entry and hence trigger the revalidation logic for the forced cache case. It would be be helpful to add a note about this for reference.

Reply:
Added comments